modules/gateway/peers.go

```diff
@@ -233,37 +233,29 @@ func (g *Gateway) Disconnect(addr modules.NetAddress) error {
 	return nil
 }
 
-// makeOutboundConnections tries to keep the Gateway well-connected. As long
-// as the Gateway is not well-connected, it tries to add random nodes as
-// peers. It sleeps when the Gateway becomes well-connected, or it has tried
-// more than 100 nodes.
-func (g *Gateway) makeOutboundConnections() {
+// peerManager tries to keep the Gateway well-connected. As long as the
+// Gateway is not well-connected, it tries to connect to random nodes.
+func (g *Gateway) peerManager() {
+	sleepTime := 5 * time.Second
 	for {
-		for i := 0; i < 100; i++ {
-			id := g.mu.RLock()
-			numPeers := len(g.peers)
-			addr, err := g.randomNode()
-			g.mu.RUnlock(id)
-			if err != nil || numPeers >= wellConnectedThreshold {
-				break
-			}
-			err = g.Connect(addr)
-			// aggressively remove unresponsive nodes
-			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
-				id = g.mu.Lock()
-				g.removeNode(addr)
-				g.mu.Unlock(id)
-			}
-		}
-		// request more nodes if necessary
+		time.Sleep(sleepTime)
+
 		id := g.mu.RLock()
-		numNodes := len(g.nodes)
-		addr, err := g.randomPeer()
+		numPeers := len(g.peers)
+		addr, err := g.randomNode()
 		g.mu.RUnlock(id)
-		if build.Release != "testing" && err == nil && numNodes < minNodeListLen {
-			g.RPC(addr, "ShareNodes", g.requestNodes)
+		if err != nil {
+			// can't do much until we have nodes
+			continue
+		}
+
+		if numPeers >= wellConnectedThreshold {
+			sleepTime = time.Minute
+		} else {
+			sleepTime = 5 * time.Second
 		}
-		time.Sleep(5 * time.Second)
+
+		g.Connect(addr)
```

Comment:
how do you feel about

`go g.Connect(addr)` ?

Then, if a peer takes 2 minutes to respond, you are still making other connections. Worst case you end up with 40 or so peers because they are all taking 2 minutes before being added. But that's still well below the 128 limit, and a highly unlikely in a real situation. Even in the black swan case it's still okay.